### PR TITLE
Implemented list projects for GitServices (#485)

### PR DESCRIPTION
**Pull request by abkosar**
Implemented list projects for GitServices (#485 )

@lachmanfrantisek

### CODE REVIEW

File: ogr/abstract.py [hunk 1]
```diff
@@ -806,6 +806,22 @@ def project_create(
         """
         raise NotImplementedError()
 
+    def list_projects(
+        self,
+        namespace: str = None,
+        user: str = None,
+        search_patter: str = None,
```
Comment (lachmanfrantisek) [inline ogr/abstract.py:813]: Probably a typo.

```suggestion
        search_pattern: str = None,
```

Would be nice to mention that the pattern is for project name -- I would cover it in the attribute name or in a docsring.

File: ogr/services/github/service.py [hunk 2]
```diff
@@ -178,3 +179,41 @@ def project_create(
...
+
+        # Get the locals for the function
+        passed_args = locals()
+
+        # Get the arguments which are not None and remove self and g
+        args_for_query = {
+            k: v
+            for k, v in passed_args.items()
+            if v and k not in ["self", "g", "search_patter"]
+        }
```
Comment (lachmanfrantisek) [inline ogr/services/github/service.py:200]: I would try to be explicit and build the query manually. This looks very nice but it's maybe harder to read.

(And `mypy` would like the explicitness as well.)

File: ogr/services/github/service.py [hunk 3]
```diff
@@ -178,3 +179,41 @@ def project_create(
...
         return PyGithubInstance(login_or_token=token)
+
+    def list_projects(
+        self,
+        namespace: str = None,
+        user: str = None,
+        search_patter: str = None,
+        language: str = None,
+    ) -> List[GitProject]:
+        g = self.github
```
Comment (lachmanfrantisek) [inline ogr/services/github/service.py:190]: You don't need to save it to a variable. It's a lazy property that is created only when needed and `g` is not a good name for a variable. I would say we don't get anything from this variable.

File: ogr/services/github/service.py (hunk 2 above)
Comment (lachmanfrantisek) [reply ogr/services/github/service.py:200]: Like you've done in case of GitLab.

File: ogr/services/gitlab/service.py [hunk 4]
```diff
@@ -123,3 +123,32 @@ def project_create(
...
+        language: str = None,
+    ) -> List[GitProject]:
+        gl = self.gitlab_instance
+
+        if namespace:
+            group = gl.groups.get(namespace)
+            projects = group.projects.list()
+        if user:
+            u = gl.users.list(username=user)[0]
+            projects = u.projects.list()
```
Comment (lachmanfrantisek) [inline ogr/services/gitlab/service.py:141]: I know that I am sometimes really strict about the code style..;) Can you please use the full name for variables? (The code is more read than written.)

And as I wrote above for `self.github`, you don't need to save `self.gitlab_instance` to a variable.

**Review by lachmanfrantisek — COMMENTED** [review]
Thanks a lot for your contribution! Looks good so far!

* I've put a few notes about the style.
* Please, don't remove the response files of other tests. (The three YAML files you've deleted.)
* There are a lot of commits created by the merges you've done. Can you please `squash` them to a single one? (I would suggest using `git rebase` next time that can apply your commits on top of the current master.)
* It would be nice to have some tests for this.

Let me know if you need some help with anything from the list above.

And btw, it looks like mypy does not complain about the type annotations, nice!

**Comment by abkosar** [comment]
@lachmanfrantisek Thanks for the comments! And I am good with you being strict about code style, I'm here to learn so no problem on that end:) Thanks for being patient about this and allowing me to work on it at my pace.

I started working on your comments. One thing I wanted to mention though. Those three .yaml files, I didn't delete them (on purpose at least). I don't know what happened but I have a guess. Every time I start working on the project, in order to keep my fork up to date I do:
- `git fetch upstream`
- `git merge upstream/master`

I think those files got deleted one of those times. Is there a way to remove them from that PR?

**Comment by abkosar** [comment]
Also, since I created the PR, is it safe to squash them now? I don't want to mess up the PR. After completing all your comments, should do take all the past 12-13 commits and do `git rebase -i HEAD~13` or is there a better way to do it?

**Comment by lachmanfrantisek** [comment]
> Also, since I created the PR, is it safe to squash them now?

Yes, sure. It's not a problem to overwrite history in your development branch. Ideally, we want the brach to be prepared and in a form to be fast-fordable on final-merge. (So you can use `rebase` instead of `merge` to achieve this.) We have some info about that [here](https://github.com/packit/contributing#git-workflow).

An example from packit repo. To have a clean git history, we want something like this:
Screenshot from 2021-01-26 10-02-56

Not this:
Screenshot from 2021-01-26 10-04-51
or this:
Screenshot from 2021-01-26 10-03-55

**Comment by lachmanfrantisek** [comment]
>  I don't want to mess up the PR. 

This will do the opposite. The final git tree will look much nicer. (And if you mess something, you will learn some git at least and you can create a new pull-request in the worst case.)

**Comment by lachmanfrantisek** [comment]
> I think those files got deleted one of those times. Is there a way to remove them from that PR?

It looks like they are removed in the [very first commit](https://github.com/packit/ogr/pull/522/commits/7da3851142cbeeadd752416e351062475eb1bd2d).

But you can still revert that:
* `git checkout master -- tests/integration/github/test_data`
  * This should checkout the version of that directory from master.
* Then, you can add those files and commit.
* Finally, to have a clean history, you can use interactive revert to squash/reorder the commits so this change dissapear.

**Comment by abkosar** [comment]
@lachmanfrantisek Thanks for all the feedback, I appreciate it!

I started working on them, work has been busy this week. As soon as I complete the code fixes, I will take care of the git part and hopefully I won't have to do another PR:)

**Comment by abkosar** [comment]
@lachmanfrantisek 

I completed all of the refactors and now I'm at the re-adding those 3 yaml files.

However I tried couple of things listed below to checkout to that version of that folder and they didn't work:
- ` git checkout master -- tests/integration/github/test_data`
- `git checkout <that_commit_hash> -- tests/integration/github/test_data`

When I do either of these nothing seems to be changing. Is there any other way to do it?

**Comment by lachmanfrantisek** [comment]
@abkosar Hmm...

- Maybe, you can try specifying the full path to each file (`git checkout master --  tests/integration/github/test_data/something.yaml`).
- You can try resetting the whole tree by `git reset master` and committing the selected three files.
  - Note that `git reset` will bring you the master version to your working space and actually reverts all your work. (But you can commit what you need and use `git reset --hard HEAD` to get rid of other changes in your working directory.)
- You can always checkout master and make a copy of those files, but that's definitely not the git way of dealing with this...;)


In general, it's good to search more about some git commands (e.g. [this](https://www.freecodecamp.org/news/save-the-day-with-git-reset/) I've found relevant) in such situations. It's worth that time.

**Comment by abkosar** [comment]
@lachmanfrantisek Thanks for that article! It was super helpful and I think he has a series which I am definitely going to read.

Ok so here is the status. I was able to recover those files and squash my commits however when i did a final `git push` to my fork I got the following:
```
black....................................................................Passed
prettier.................................................................Passed
Check for added large files..............................................Passed
Check python ast.........................................................Passed
Check for merge conflicts................................................Passed
Check Yaml...............................................................Passed
Detect Private Key.......................................................Passed
Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
flake8...................................................................Passed
mypy.....................................................................Passed
Check if rebase is needed................................................Failed
- hook id: check-rebase
- exit code: 1

Last commit subject: Completed list_projects functionality

Upstream hash: 5e0d8c2280cb3eb31ebc6ef12598efa19c0dd28e
Local hashes: ['8035ad6653915ec427178867670c45ca2403a175', '0c6992b8772f4b81c50d3d1ed84520ab7794c0ee', 'dd05b51b080b2d7ce42bd6f2426cc392c7a34a4f']

Your branch is not up to date with upstream/master. 
SHA of the last commit of upstream/master: 5e0d8c2280cb3eb31ebc6ef12598efa19c0dd28e
Please, rebase! 

Last commit subject: Completed list_projects functionality

Upstream hash: 5e0d8c2280cb3eb31ebc6ef12598efa19c0dd28e
Local hashes: ['8035ad6653915ec427178867670c45ca2403a175', '0c6992b8772f4b81c50d3d1ed84520ab7794c0ee', 'dd05b51b080b2d7ce42bd6f2426cc392c7a34a4f']

Your branch is not up to date with upstream/master. 
SHA of the last commit of upstream/master: 5e0d8c2280cb3eb31ebc6ef12598efa19c0dd28e
Please, rebase!

Requre response files cleanup............................................Passed
error: failed to push some refs to 'git@github.com:abkosar/ogr.git'
```
I also did a `git rebase upstream/master` however it didn't work either. Do you have any suggestions?

Thanks!

**Comment by csomh** [comment]
@abkosar Please be aware that we've renamed our default branch some time ago to `main`.

Try to run `git fetch --all` to get all the changes from all the remotes you have configured and then `git rebase upstream/main` from your feature branch, to rebase the feature branch.

You might also want to rename your local default branch to `main` (with `git branch -m main`) and set up tracking for that if needed (with `git branch --set-upstream-to=upstream/main`, or similar, depending which remote you want to track).

**Comment by abkosar** [comment]
@csomh Thanks for the advice it was very helpful.

I completed everything, now I have a pretty clean PR however since I changed my branch name it didn't update this PR so is it ok I create a new one which looks much cleaner then this one?

**Comment by csomh** [comment]
@abkosar sure, you can do that, though I see that the branch called "master", from which you've opened this PR, is still available in your fork (that is, on GitHub). You can simply try to force-push your local changes there (you are rewriting history after a rebase, hence the "force" in push :slightly_smiling_face: ), which will update this PR. After you've checked out the branch you did you're work on (you can check this with a `git log` to see if you have the commits you want), run `git push --force-with-lease origin master`.

**Comment by csomh** [comment]
Nevermind, I see you've already opened #538 . So closing this, to keep things tight and clean.

**Comment by abkosar** [comment]
Sounds good, yeah I saw your comment after I opened the PR. Let's take it from there, thanks!

**Pull request closed without merging**